File: src/corral/agents/claude.py
"""Claude agent implementation."""

from __future__ import annotations

import json
import re
import shutil
from pathlib import Path
from typing import Any

from corral.agents.base import BaseAgent
from corral.utils import HISTORY_PATH

SUMMARY_RE = re.compile(r"^[\s\u25cf\u23fa]*\|\|PULSE:SUMMARY (.*?)\|\|", re.MULTILINE)

FTS_BODY_CAP = 50_000


def _clean_match(text: str) -> str:
    return " ".join(text.split())


def _hook_entry_exists(matcher_groups: list, command: str) -> bool:
    """Check if a hook command already exists in a list of matcher groups."""
    for group in matcher_groups:
        for hook in group.get("hooks", []):
            if hook.get("command") == command:
                return True
    return False


def install_hooks(target_dir: Path | str | None = None):
    """Ensure Claude hooks for Corral are installed in settings.local.json.

    When *target_dir* is provided, hooks are written to
    ``<target_dir>/.claude/settings.local.json`` (per-worktree).
    When omitted, falls back to ``~/.claude/settings.local.json`` (global).

    Merges corral hooks into existing user hooks without overriding them.
    Skips any hook that is already present (matched by command name).
    """
    if target_dir is not None:
        target = Path(target_dir)
        # Only install into git worktree roots (.git file or directory)
        if not (target / ".git").exists():
            return
        settings_path = target / ".claude" / "settings.local.json"
    else:
        settings_path = Path.home() / ".claude" / "settings.local.json"
    settings_path.parent.mkdir(parents=True, exist_ok=True)

    if settings_path.exists():
        settings = json.loads(settings_path.read_text())
    else:
        settings = {}

    # Clean up any old-format keys from previous versions
    for old_key in ("agenticStateHook", "taskStateHook"):
        settings.pop(old_key, None)

    hooks = settings.setdefault("hooks", {})

    # Hooks we want to ensure exist: (event, matcher_group_to_add)
    desired = [
        ("PostToolUse", {
            "matcher": "TaskCreate|TaskUpdate",
            "hooks": [{"type": "command", "command": "corral-hook-task-sync"}],
        }),
        ("PostToolUse", {
            "hooks": [{"type": "command", "command": "corral-hook-agentic-state"}],
        }),
        ("Stop", {
            "hooks": [{"type": "command", "command": "corral-hook-agentic-state"}],
        }),
        ("Notification", {
            "hooks": [{"type": "command", "command": "corral-hook-agentic-state"}],
        }),
    ]

    modified = False
    for event, group in desired:
        event_list = hooks.setdefault(event, [])
        command = group["hooks"][0]["command"]
        if not _hook_entry_exists(event_list, command):
            event_list.append(group)
            modified = True

    if modified:
        settings_path.write_text(json.dumps(settings, indent=2) + "\n")


def _extract_text_from_entry(entry: dict) -> str:
    """Extract plain text from a Claude JSONL message entry."""
    msg = entry.get("message", {})
    content = msg.get("content", "")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return "\n".join(
            b.get("text", "")
            for b in content
            if isinstance(b, dict) and b.get("type") == "text"
        )
    return ""


class ClaudeAgent(BaseAgent):
    """Claude Code agent."""

    @property
    def agent_type(self) -> str:
        return "claude"

    @property
    def supports_resume(self) -> bool:
        return True

    @property
    def history_base_path(self) -> Path:
        return HISTORY_PATH

    @property
    def history_glob_pattern(self) -> str:
        return "*.jsonl"

    def build_launch_command(
        self,
        session_id: str,
        protocol_path: Path | None,
        resume_session_id: str | None = None,
        flags: list[str] | None = None,
    ) -> str:
        parts = ["claude"]
        if resume_session_id:
            parts.append(f"--resume {resume_session_id}")
        else:
            parts.append(f"--session-id {session_id}")
        if protocol_path and protocol_path.exists():
            parts.append(f"--append-system-prompt \"$(cat '{protocol_path}')\"")
        if flags:
            parts.extend(flags)
        return " ".join(parts)

    def install_hooks(self, working_dir: str) -> None:
        install_hooks(working_dir)  # module-level function defined above

    def prepare_resume(self, session_id: str, working_dir: str) -> None:
        """Copy session JSONL into the target project dir so ``claude --resume`` finds it."""
        encoded = working_dir.replace("/", "-").replace("_", "-")
        target_project = HISTORY_PATH / encoded
        target_jsonl = target_project / f"{session_id}.jsonl"

        if target_jsonl.exists():
            return

        source_jsonl: Path | None = None
        for candidate in HISTORY_PATH.iterdir():
            if not candidate.is_dir():
                continue
            f = candidate / f"{session_id}.jsonl"
            if f.exists():
                source_jsonl = f
                break

        if source_jsonl is None:
            return

        target_project.mkdir(parents=True, exist_ok=True)
        shutil.copy2(source_jsonl, target_jsonl)

        source_dir = source_jsonl.parent / session_id
        target_dir = target_project / session_id
        if source_dir.is_dir() and not target_dir.exists():
            shutil.copytree(source_dir, target_dir)

    def load_history_sessions(self) -> list[dict[str, Any]]:
        history_base = Path.home() / ".claude" / "projects"
        if not history_base.exists():
            return []

        sessions: dict[str, dict[str, Any]] = {}
        for history_file in history_base.rglob("*.jsonl"):
            try:
                with open(history_file, "r", errors="replace") as f:
                    for line in f:
                        line = line.strip()
                        if not line:
                            continue
                        try:
                            entry = json.loads(line)
                        except json.JSONDecodeError:
                            continue

                        session_id = entry.get("sessionId")
                        if not session_id:
                            continue

                        if session_id not in sessions:
                            sessions[session_id] = {
                                "session_id": session_id,
                                "messages": [],
                                "first_timestamp": entry.get("timestamp"),
                                "last_timestamp": entry.get("timestamp"),
                                "source_file": str(history_file),
                                "source_type": "claude",
                                "summary": None,
                            }

                        ts = entry.get("timestamp")
                        if ts:
                            if not sessions[session_id]["first_timestamp"] or ts < sessions[session_id]["first_timestamp"]:
                                sessions[session_id]["first_timestamp"] = ts
                            if not sessions[session_id]["last_timestamp"] or ts > sessions[session_id]["last_timestamp"]:
                                sessions[session_id]["last_timestamp"] = ts

                        sessions[session_id]["messages"].append(entry)
            except OSError:
                continue

        result = []
        for sid, data in sessions.items():
            summary_marker = ""
            first_human = ""
            for msg in data["messages"]:
                if not summary_marker and msg.get("type") == "assistant":
                    content = msg.get("message", {}).get("content", "")
                    text = ""
                    if isinstance(content, str):
                        text = content
                    elif isinstance(content, list):
                        text = " ".join(
                            b.get("text", "") for b in content
                            if isinstance(b, dict) and b.get("type") == "text"
                        )
                    m = SUMMARY_RE.search(text)
                    if m:
                        summary_marker = _clean_match(m.group(1))

                if not first_human and msg.get("type") in ("human", "user"):
                    content = msg.get("message", {}).get("content", "")
                    if isinstance(content, str):
                        first_human = content[:100]
                    elif isinstance(content, list):
                        for block in content:
                            if isinstance(block, dict) and block.get("type") == "text":
                                first_human = block.get("text", "")[:100]
                                break
            data["summary"] = summary_marker or first_human or "(no messages)"
            data["message_count"] = len(data["messages"])
            listing = {k: v for k, v in data.items() if k != "messages"}
            result.append(listing)

        return result

    def load_session_messages(self, session_id: str) -> list[dict[str, Any]]:
        history_base = Path.home() / ".claude" / "projects"
        if not history_base.exists():
            return []

        messages = []
        for history_file in history_base.rglob("*.jsonl"):
            try:
                with open(history_file, "r", errors="replace") as f:
                    for line in f:
                        line = line.strip()
                        if not line:
                            continue
                        try:
                            entry = json.loads(line)
                        except json.JSONDecodeError:
                            continue
                        if entry.get("sessionId") == session_id:
                            messages.append(entry)
            except OSError:
                continue

        return messages

    async def index_file(self, path: Path, mtime: float, store: Any) -> int:
        """Parse a JSONL file and upsert each session found."""
        from corral.session_manager import SUMMARY_RE as SM_SUMMARY_RE, clean_match

        sessions: dict[str, dict[str, Any]] = {}
        try:
            with open(path, "r", errors="replace") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        entry = json.loads(line)
                    except json.JSONDecodeError:
                        continue

                    sid = entry.get("sessionId")
                    if not sid:
                        continue

                    if sid not in sessions:
                        sessions[sid] = {
                            "texts": [],
                            "first_ts": entry.get("timestamp"),
                            "last_ts": entry.get("timestamp"),
                            "msg_count": 0,
                            "summary_marker": "",
                            "first_human": "",
                        }

                    s = sessions[sid]
                    s["msg_count"] += 1
                    ts = entry.get("timestamp")
                    if ts:
                        if not s["first_ts"] or ts < s["first_ts"]:
                            s["first_ts"] = ts
                        if not s["last_ts"] or ts > s["last_ts"]:
                            s["last_ts"] = ts

                    text = _extract_text_from_entry(entry)
                    if text.strip():
                        s["texts"].append(text)

                    if not s["summary_marker"] and entry.get("type") == "assistant":
                        m = SM_SUMMARY_RE.search(text)
                        if m:
                            s["summary_marker"] = clean_match(m.group(1))

                    if not s["first_human"] and entry.get("type") in ("human", "user"):
                        s["first_human"] = text[:100]
        except OSError:
            return 0

        for sid, s in sessions.items():
            summary = s["summary_marker"] or s["first_human"] or "(no messages)"
            await store.upsert_session_index(
                session_id=sid,
                source_type="claude",
                source_file=str(path),
                first_timestamp=s["first_ts"],
                last_timestamp=s["last_ts"],
                message_count=s["msg_count"],
                display_summary=summary,
                file_mtime=mtime,
            )
            body = "\n".join(s["texts"])[:FTS_BODY_CAP]
            await store.upsert_fts(sid, body)
            await store.enqueue_for_summarization(sid)

        return len(sessions)

File: coral-go/internal/agent/claude.go
package agent

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
)

// ClaudeAgent implements the Agent interface for Claude Code.
type ClaudeAgent struct{}

func (a *ClaudeAgent) AgentType() string    { return "claude" }
func (a *ClaudeAgent) SupportsResume() bool { return true }

func (a *ClaudeAgent) HistoryBasePath() string {
	if v := os.Getenv("CLAUDE_PROJECTS_DIR"); v != "" {
		return v
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".claude", "projects")
}

func (a *ClaudeAgent) HistoryGlobPattern() string { return "*.jsonl" }

// ExtractSessions scans Claude history files under basePath and returns indexed sessions.
// Files whose mtime matches knownMtimes are skipped.
func (a *ClaudeAgent) ExtractSessions(basePath string, knownMtimes map[string]float64) ([]IndexedSession, error) {
	if basePath == "" {
		return nil, nil
	}
	entries, err := os.ReadDir(basePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var sessions []IndexedSession
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		projectDir := filepath.Join(basePath, entry.Name())
		files, err := filepath.Glob(filepath.Join(projectDir, a.HistoryGlobPattern()))
		if err != nil {
			continue
		}
		for _, fpath := range files {
			info, err := os.Stat(fpath)
			if err != nil {
				continue
			}
			mtime := float64(info.ModTime().Unix())
			if prev, ok := knownMtimes[fpath]; ok && prev == mtime {
				continue // file unchanged since last index
			}
			sess, err := parseClaudeSessions(fpath, mtime)
			if err != nil {
				slog.Debug("claude: failed to parse session file", "path", fpath, "error", err)
				continue
			}
			sessions = append(sessions, sess...)
		}
	}
	return sessions, nil
}

// parseClaudeSessions parses a Claude JSONL file and extracts session metadata.
func parseClaudeSessions(fpath string, mtime float64) ([]IndexedSession, error) {
	f, err := os.Open(fpath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	sessionID := strings.TrimSuffix(filepath.Base(fpath), ".jsonl")
	var firstTS, lastTS *string
	var msgCount int
	var summaryParts []string

	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var entry map[string]any
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}
		ts, _ := entry["timestamp"].(string)
		if ts != "" {
			if firstTS == nil {
				firstTS = &ts
			}
			lastCopy := ts
			lastTS = &lastCopy
		}
		etype, _ := entry["type"].(string)
		if etype == "user" || etype == "assistant" {
			msgCount++
		}
		// Grab first assistant text for display summary
		if etype == "assistant" && len(summaryParts) < 1 {
			if msg, _ := entry["message"].(map[string]any); msg != nil {
				if text := extractFirstText(msg["content"]); text != "" {
					if len(text) > 200 {
						text = text[:200]
					}
					summaryParts = append(summaryParts, text)
				}
			}
		}
	}
	if msgCount == 0 {
		return nil, nil
	}
	return []IndexedSession{{
		SessionID:      sessionID,
		SourceType:     "claude",
		SourceFile:     fpath,
		FileMtime:      mtime,
		FirstTimestamp: firstTS,
		LastTimestamp:   lastTS,
		MessageCount:   msgCount,
		DisplaySummary: strings.Join(summaryParts, " "),
	}}, nil
}

// extractFirstText gets the first text string from message content.
func extractFirstText(content any) string {
	switch c := content.(type) {
	case string:
		return c
	case []any:
		for _, block := range c {
			if b, ok := block.(map[string]any); ok {
				if bt, _ := b["type"].(string); bt == "text" {
					if t, _ := b["text"].(string); t != "" {
						return t
					}
				}
			}
		}
	}
	return ""
}

func (a *ClaudeAgent) BuildLaunchCommand(params LaunchParams) string {
	bin := resolveBinary(params.CLIPath, "claude")
	parts := []string{bin}

	effectiveID := params.SessionID
	if params.ResumeSessionID != "" {
		effectiveID = params.ResumeSessionID
		parts = append(parts, "--resume", params.ResumeSessionID)
	} else {
		parts = append(parts, "--session-id", params.SessionID)
	}

	// Build merged settings with hooks and system prompt
	merged := buildMergedSettings(params.WorkingDir, params.Hooks)

	// Combine protocol + board system prompt into systemPrompt
	var sysParts []string
	if proto := readProtocolFile(params.ProtocolPath); proto != "" {
		sysParts = append(sysParts, proto)
	}
	boardSysPrompt := BuildBoardSystemPrompt(params.BoardName, params.Role, params.Prompt, params.PromptOverrides, params.BoardType)
	if boardSysPrompt != "" {
		sysParts = append(sysParts, boardSysPrompt)
	}
	if len(sysParts) > 0 {
		merged["systemPrompt"] = strings.Join(sysParts, "\n\n")
	}

	// Inject permissions from capabilities
	if perms := TranslateToClaudePermissions(params.Capabilities); perms != nil {
		permMap := map[string]any{}
		if len(perms.Allow) > 0 {
			permMap["allow"] = perms.Allow
		}
		if len(perms.Deny) > 0 {
			permMap["deny"] = perms.Deny
		}
		merged["permissions"] = permMap
	}

	// Inject allowed tools
	if len(params.Tools) > 0 {
		merged["allowedTools"] = params.Tools
	}

	// Inject MCP server configs
	if len(params.MCPServers) > 0 {
		merged["mcpServers"] = params.MCPServers
	}

	// Set CORAL_SESSION_NAME and CORAL_SUBSCRIBER_ID in env so coral-board and hooks
	// can identify this agent. CORAL_SUBSCRIBER_ID is the stable board identity (role name).
	{
		envMap, _ := merged["env"].(map[string]interface{})
		if envMap == nil {
			envMap = make(map[string]interface{})
		}
		if params.SessionName != "" {
			envMap["CORAL_SESSION_NAME"] = params.SessionName
		}
		if params.Role != "" {
			envMap["CORAL_SUBSCRIBER_ID"] = params.Role
		}
		if params.ProxyBaseURL != "" {
			// Override the appropriate base URL env var based on detected provider.
			// This reroutes the Claude agent through our proxy regardless of provider.
			// Note: Claude CLI only uses Anthropic providers (direct, Bedrock, Vertex).
			// OpenAI/Codex agents handle their own proxy URL override separately.
			switch params.UpstreamProvider {
			case "bedrock":
				envMap["ANTHROPIC_BEDROCK_BASE_URL"] = params.ProxyBaseURL
			case "vertex":
				envMap["ANTHROPIC_VERTEX_BASE_URL"] = params.ProxyBaseURL
			default:
				envMap["ANTHROPIC_BASE_URL"] = params.ProxyBaseURL
			}
		}
		merged["env"] = envMap
	}

	// Add Coral tools dir to PATH in env settings so that coral-board,
	// coral hooks, etc. can be found by Claude CLI and its subprocesses.
	if macosDir := CoralToolsDir(); macosDir != "" {
		envMap, _ := merged["env"].(map[string]interface{})
		if envMap == nil {
			envMap = make(map[string]interface{})
		}
		if existingPath, ok := envMap["PATH"].(string); ok {
			// Prepend if not already present
			if !strings.Contains(existingPath, macosDir) {
				envMap["PATH"] = macosDir + ":" + existingPath
			}
		} else {
			// Resolve $PATH from the actual environment — JSON settings files
			// are not shell-expanded, so literal "$PATH" would be passed as-is.
			currentPath := os.Getenv("PATH")
			envMap["PATH"] = macosDir + ":" + currentPath
		}
		merged["env"] = envMap
	}

	// Write settings to temp file using writeTempFile for safe creation
	data, _ := json.MarshalIndent(merged, "", "  ")
	settingsFile := writeTempFile("settings", effectiveID, "json", append(data, '\n'))
	parts = append(parts, "--settings", settingsFile)

	if params.PermissionMode != "" && params.PermissionMode != "default" {
		parts = append(parts, "--permission-mode", params.PermissionMode)
	}

	if len(params.Flags) > 0 {
		parts = append(parts, params.Flags...)
	}

	// Pass the prompt as a CLI positional argument so the agent starts immediately
	// without relying on fragile tmux send-keys delivery.
	cliPrompt := BuildBoardActionPrompt(params.BoardName, params.Role, params.Prompt, params.PromptOverrides, params.BoardType)
	if cliPrompt != "" {
		promptFile := writeTempFile("prompt", effectiveID, "txt", []byte(cliPrompt))
		parts = append(parts, FormatPromptFileArg(promptFile))
	}

	return strings.Join(ShellQuoteParts(parts), " ")
}

func (a *ClaudeAgent) PrepareResume(sessionID, workingDir string) {
	basePath := a.HistoryBasePath()
	encoded := strings.ReplaceAll(strings.ReplaceAll(workingDir, "/", "-"), "_", "-")
	targetProject := filepath.Join(basePath, encoded)
	targetJSONL := filepath.Join(targetProject, sessionID+".jsonl")

	if _, err := os.Stat(targetJSONL); err == nil {
		return // Already exists
	}

	// Search for the session file in other project dirs
	entries, err := os.ReadDir(basePath)
	if err != nil {
		return
	}
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		source := filepath.Join(basePath, entry.Name(), sessionID+".jsonl")
		if _, err := os.Stat(source); err == nil {
			os.MkdirAll(targetProject, 0755)
			data, err := os.ReadFile(source)
			if err == nil {
				os.WriteFile(targetJSONL, data, 0644)
			}
			// Copy session directory if it exists
			sourceDir := filepath.Join(basePath, entry.Name(), sessionID)
			targetDir := filepath.Join(targetProject, sessionID)
			if info, err := os.Stat(sourceDir); err == nil && info.IsDir() {
				copyDir(sourceDir, targetDir)
			}
			return
		}
	}
}

// Coral hooks to inject into every Claude session.
var coralHooks = map[string][]map[string]interface{}{
	"PostToolUse": {
		{"matcher": "TaskCreate|TaskUpdate", "hooks": []map[string]interface{}{
			{"type": "command", "command": "coral-hook-task-sync"},
		}},
		{"hooks": []map[string]interface{}{
			{"type": "command", "command": "coral-hook-agentic-state"},
		}},
		{"hooks": []map[string]interface{}{
			{"type": "command", "command": "coral-hook-message-check"},
		}},
	},
	"Stop": {
		{"hooks": []map[string]interface{}{
			{"type": "command", "command": "coral-hook-agentic-state"},
		}},
	},
	"Notification": {
		{"hooks": []map[string]interface{}{
			{"type": "command", "command": "coral-hook-agentic-state"},
		}},
	},
	"SessionStart": {
		{"hooks": []map[string]interface{}{
			{"type": "command", "command": "coral-hook-agentic-state"},
		}},
		{"hooks": []map[string]interface{}{
			{"type": "command", "command": "coral-hook-session-start"},
		}},
	},
}

func buildMergedSettings(workingDir string, agentHooks map[string]interface{}) map[string]interface{} {
	home, _ := os.UserHomeDir()
	homeClaude := filepath.Join(home, ".claude")

	global := readSettingsFile(filepath.Join(homeClaude, "settings.json"))
	var project, local map[string]interface{}
	if workingDir != "" {
		project = readSettingsFile(filepath.Join(workingDir, ".claude", "settings.json"))
		local = readSettingsFile(filepath.Join(workingDir, ".claude", "settings.local.json"))
	}

	// Shallow merge: local > project > global
	merged := make(map[string]interface{})
	for k, v := range global {
		merged[k] = v
	}
	for k, v := range project {
		merged[k] = v
	}
	for k, v := range local {
		merged[k] = v
	}

	// Deep-merge env maps (project env should not replace global env entirely)
	mergedEnv := make(map[string]interface{})
	for _, source := range []map[string]interface{}{global, project, local} {
		if env, ok := source["env"].(map[string]interface{}); ok {
			for k, v := range env {
				mergedEnv[k] = v
			}
		}
	}
	if len(mergedEnv) > 0 {
		merged["env"] = mergedEnv
	}

	// Deep-merge hooks
	mergedHooks := make(map[string][]interface{})
	for _, source := range []map[string]interface{}{global, project, local} {
		if hooks, ok := source["hooks"].(map[string]interface{}); ok {
			for event, groups := range hooks {
				if groupList, ok := groups.([]interface{}); ok {
					mergedHooks[event] = append(mergedHooks[event], groupList...)
				}
			}
		}
	}

	// Append Coral hooks
	for event, groups := range coralHooks {
		existing := mergedHooks[event]
		for _, group := range groups {
			command := ""
			if hooks, ok := group["hooks"].([]map[string]interface{}); ok && len(hooks) > 0 {
				if cmd, ok := hooks[0]["command"].(string); ok {
					command = cmd
				}
			}
			if !hookEntryExists(existing, command) {
				mergedHooks[event] = append(mergedHooks[event], group)
			}
		}
	}

	// Append per-agent hooks (level 5: workflow step / team agent config)
	for event, groups := range agentHooks {
		if groupList, ok := groups.([]interface{}); ok {
			mergedHooks[event] = append(mergedHooks[event], groupList...)
		}
	}

	merged["hooks"] = mergedHooks
	return merged
}

func readSettingsFile(path string) map[string]interface{} {
	data, err := os.ReadFile(path)
	if err != nil {
		return map[string]interface{}{}
	}
	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return map[string]interface{}{}
	}
	return result
}

func hookEntryExists(groups []interface{}, command string) bool {
	if command == "" {
		return false
	}
	for _, g := range groups {
		if gMap, ok := g.(map[string]interface{}); ok {
			if hooks, ok := gMap["hooks"].([]interface{}); ok {
				for _, h := range hooks {
					if hMap, ok := h.(map[string]interface{}); ok {
						if hMap["command"] == command {
							return true
						}
					}
				}
			}
			// Also handle typed version
			if hooks, ok := gMap["hooks"].([]map[string]interface{}); ok {
				for _, h := range hooks {
					if h["command"] == command {
						return true
					}
				}
			}
		}
	}
	return false
}


// BuildMergedSettingsForDetection returns the deep-merged env map from user settings files.
// This is used by the launcher to detect the upstream provider before building the full
// launch command.
func BuildMergedSettingsForDetection(workingDir string) map[string]interface{} {
	home, _ := os.UserHomeDir()
	homeClaude := filepath.Join(home, ".claude")

	global := readSettingsFile(filepath.Join(homeClaude, "settings.json"))
	var project, local map[string]interface{}
	if workingDir != "" {
		project = readSettingsFile(filepath.Join(workingDir, ".claude", "settings.json"))
		local = readSettingsFile(filepath.Join(workingDir, ".claude", "settings.local.json"))
	}

	// Deep-merge env maps
	mergedEnv := make(map[string]interface{})
	for _, source := range []map[string]interface{}{global, project, local} {
		if env, ok := source["env"].(map[string]interface{}); ok {
			for k, v := range env {
				mergedEnv[k] = v
			}
		}
	}
	return mergedEnv
}

// UpstreamInfo holds the detected upstream provider and URL for proxy rerouting.
type UpstreamInfo struct {
	Provider    string // "anthropic", "bedrock", "vertex", "openai"
	UpstreamURL string // real endpoint before proxy override
}

// DetectUpstreamURL inspects the merged env block and OS env vars to determine
// which upstream LLM provider the agent is configured to use. Returns the
// provider type and the real upstream URL.
//
// Detection priority: merged settings env > OS env > defaults.
// Check order: Bedrock (most specific) > Vertex > Direct Anthropic > OpenAI > default.
func DetectUpstreamURL(mergedEnv map[string]interface{}) UpstreamInfo {
	// Check for Bedrock first (more specific)
	if url := envOrOS(mergedEnv, "ANTHROPIC_BEDROCK_BASE_URL"); url != "" {
		return UpstreamInfo{Provider: "bedrock", UpstreamURL: url}
	}
	if envOrOS(mergedEnv, "CLAUDE_CODE_USE_BEDROCK") == "1" {
		region := envOrOS(mergedEnv, "AWS_REGION")
		if region == "" {
			region = "us-east-1"
		}
		return UpstreamInfo{
			Provider:    "bedrock",
			UpstreamURL: fmt.Sprintf("https://bedrock-runtime.%s.amazonaws.com", region),
		}
	}

	// Check for Vertex AI
	if url := envOrOS(mergedEnv, "ANTHROPIC_VERTEX_BASE_URL"); url != "" {
		return UpstreamInfo{Provider: "vertex", UpstreamURL: url}
	}
	if envOrOS(mergedEnv, "CLAUDE_CODE_USE_VERTEX") == "1" {
		// Vertex default URL requires project/location, use placeholder
		return UpstreamInfo{Provider: "vertex", UpstreamURL: ""}
	}

	// Direct Anthropic (custom base URL)
	if url := envOrOS(mergedEnv, "ANTHROPIC_BASE_URL"); url != "" {
		return UpstreamInfo{Provider: "anthropic", UpstreamURL: url}
	}

	// OpenAI
	if url := envOrOS(mergedEnv, "OPENAI_BASE_URL"); url != "" {
		return UpstreamInfo{Provider: "openai", UpstreamURL: url}
	}

	// Default: direct Anthropic API
	return UpstreamInfo{Provider: "anthropic", UpstreamURL: "https://api.anthropic.com"}
}

// envOrOS checks the merged env map first, then falls back to os.Getenv().
func envOrOS(mergedEnv map[string]interface{}, key string) string {
	if v, ok := mergedEnv[key]; ok {
		if s, ok := v.(string); ok && s != "" {
			return s
		}
	}
	return os.Getenv(key)
}

func copyDir(src, dst string) {
	entries, err := os.ReadDir(src)
	if err != nil {
		return
	}
	if err := os.MkdirAll(dst, 0755); err != nil {
		slog.Warn("copyDir: failed to create directory", "dst", dst, "error", err)
		return
	}
	for _, entry := range entries {
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, entry.Name())
		if entry.IsDir() {
			copyDir(srcPath, dstPath)
		} else {
			data, err := os.ReadFile(srcPath)
			if err != nil {
				slog.Warn("copyDir: failed to read file", "src", srcPath, "error", err)
				continue
			}
			if err := os.WriteFile(dstPath, data, 0644); err != nil {
				slog.Warn("copyDir: failed to write file", "dst", dstPath, "error", err)
			}
		}
	}
}
